segments_tree.go

```diff
@@ -0,0 +1,141 @@
+package quamina
+
+import (
+	"fmt"
+	"strings"
+)
+
+const SegmentSeparator = "\n"
+
+type segmentsTree struct {
+	root bool
+
+	nodes map[string]*segmentsTree
+
+	// fields map segment to full path
+	// will be present only on the leafs
+	fields map[string][]byte
+}
+
+func newSegmentsIndex(paths ...string) *segmentsTree {
+	st := newSegmentsIndexNode(true)
+
+	for _, path := range paths {
+		st.add(path)
+	}
+
+	return st
+}
+
+func newSegmentsIndexNode(root bool) *segmentsTree {
+	return &segmentsTree{
+		root:   root,
+		nodes:  make(map[string]*segmentsTree),
+		fields: make(map[string][]byte),
+	}
+}
+
+func (p *segmentsTree) add(path string) {
+	parts := strings.Split(path, SegmentSeparator)
+
+	if len(parts) == 1 {
+		// It's a direct field.
+		p.fields[path] = []byte(path)
+		return
+	}
+
+	var node *segmentsTree
+
+	node = p.getOrCreate(parts[0])
+
+	for i, part := range parts[1:] {
```

Comment:
This logic is baffling me.  A comment might help?

Reply:
Re-wrote the implementation and added a comment